src/trading/CompleteSets.sol

```diff
@@ -0,0 +1,73 @@
+pragma solidity ^0.4.13;
+
+import 'ROOT/trading/ICompleteSets.sol';
+import 'ROOT/Controlled.sol';
+import 'ROOT/libraries/ReentrancyGuard.sol';
+import 'ROOT/libraries/math/SafeMathUint256.sol';
+import 'ROOT/libraries/token/ERC20.sol';
+import 'ROOT/extensions/MarketFeeCalculator.sol';
+import 'ROOT/reporting/IMarket.sol';
+import 'ROOT/reporting/IReportingWindow.sol';
+import 'ROOT/trading/IOrders.sol';
+
+
+contract CompleteSets is Controlled, ReentrancyGuard, ICompleteSets {
+    using SafeMathUint256 for uint256;
+
+    /**
+     * Buys `_amount` shares of every outcome in the specified market.
+    **/
+    function publicBuyCompleteSets(IMarket _market, uint256 _amount) external onlyInGoodTimes nonReentrant returns (bool) {
+        return this.buyCompleteSets(msg.sender, _market, _amount);
+    }
+
+    function buyCompleteSets(address _sender, IMarket _market, uint256 _amount) external onlyWhitelistedCallers returns (bool) {
```

Comment:
Why make this external? Seems like private or internal is more appropriate.

Reply:
It is intended to be called by other Augur contracts (trusted callers).  `publicBuyCompleteSets` is intended to be called by untrusted external sources.  The `whitelistCallersOnly` modifier checks `msg.sender` which is why we need to call this via `this.buyCompleteSets` even when we are calling it privately because otherwise `msg.sender` will be the original `msg.sender` and the whitelisted callers check will fail.